### ISSUE TITLE
Cannot return BLOB from created function

### ISSUE THREAD

**Issue by jimflood** [sparklemotion/sqlite3-ruby#170]
There is no way to return a BLOB value from a created function (i.e. db.create_function). There is no call to sqlite3_result_blob. Returning TEXT does not work when there are embedded null bytes in the data.

I implemented a change which is to return BLOB if the encoding of the string is ASCII-8BIT and submitted a pull request. cf. the pull request.

There is a failure in the Travis integration test, but I can build and run the tests with Ruby 1.8.7 locally so I don't understand the issue there.

**Comment by jimflood** [comment]
What can I do to get this fix accepted and merged? Should I prepare a test case to show the issue, and a test to show that it is fixed?

For example:

```ruby
require 'sqlite3'
require 'base64'

db = SQLite3::Database.new("test.db");

db.create_function("echo", 1) do |func, str|
    func.result = str.force_encoding('ASCII-8BIT') # return BLOB -- may have embedded null bytes
end

text = "hello\u0000world"
puts db.execute("select ?", text)[0][0].bytes.inspect
puts db.execute("select echo(?)", text)[0][0].bytes.inspect
puts db.execute("select echo(echo(?))", text)[0][0].bytes.inspect

blob = "hello\u0000world".force_encoding('ASCII-8BIT')
puts db.execute("select ?", blob)[0][0].bytes.inspect
puts db.execute("select echo(?)", blob)[0][0].bytes.inspect
puts db.execute("select echo(echo(?))", blob)[0][0].bytes.inspect

db.close
```

The broken output is:

```
[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]
[104, 101, 108, 108, 111]
[104, 101, 108, 108, 111]
[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]
[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]
[104, 101, 108, 108, 111]
```

The null byte causes the data to be truncated. It's not apparent in the trivial case, but just work with the value in the SQL a bit, and you'll see where SQLite3 is mislead because it sees the output of the user-defined function as TEXT. A simple test of feeding the result back into the function one extra time shows the problem clearly. The last line of output should still be "[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]".

**Comment by jimflood** [comment]
Also, can anybody help me with the Travis test that fails? I can't reproduce this on my own system -- I don't know how. I can run all tests in Ruby 1.8.7 from the repo.

**Comment by tenderlove** [comment]
I think this was fixed with 220c1947976770a0a12401a467899b79f1900640 and 1.8 tests were fixed with c204cb9fbf8bf99e8b239c3365108204ba2914f8

**Comment by jimflood** [comment]
Yes, I just tried the sample ruby above with the fix gem, and the output is now correct:

[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]
[104, 101, 108, 108, 111]
[104, 101, 108, 108, 111]
[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]
[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]
[104, 101, 108, 108, 111, 0, 119, 111, 114, 108, 100]
